xds/internal/client/callback.go

```diff
@@ -74,26 +74,50 @@ func (c *clientImpl) callCallback(wiu *watcherInfoWithUpdate) {
 //
 // A response can contain multiple resources. They will be parsed and put in a
 // map from resource name to the resource content.
-func (c *clientImpl) NewListeners(updates map[string]ListenerUpdate) {
+func (c *clientImpl) NewListeners(updates map[string]ListenerUpdate, metadata UpdateMetadata) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
+	if metadata.ErrState != nil {
+		// On NACK, update overall version to the NACKed resp.
+		c.ldsVersion = metadata.ErrState.Version
+		for name := range updates {
+			if _, ok := c.ldsWatchers[name]; ok {
+				// On error, keep previous version for each resource. But update
+				// status and error.
+				mdCopy := c.ldsMD[name]
+				mdCopy.ErrState = metadata.ErrState
+				mdCopy.Status = metadata.Status
+				c.ldsMD[name] = mdCopy
+				// TODO: send the NACK error to the watcher.
+			}
+		}
+		return
+	}
+
+	// If no error received, the status is ACK.
+	c.ldsVersion = metadata.Version
 	for name, update := range updates {
 		if s, ok := c.ldsWatchers[name]; ok {
+			// Only send the update if this is not an error.
 			for wi := range s {
 				wi.newUpdate(update)
 			}
 			// Sync cache.
 			c.logger.Debugf("LDS resource with name %v, value %+v added to cache", name, update)
 			c.ldsCache[name] = update
+			c.ldsMD[name] = metadata
 		}
 	}
+	// Resources not in the new update were removed by the server, so delete
+	// them.
 	for name := range c.ldsCache {
 		if _, ok := updates[name]; !ok {
-			// If resource exists in cache, but not in the new update, delete it
-			// from cache, and also send an resource not found error to indicate
-			// resource removed.
+			// If resource exists in cache, but not in the new update, delete
+			// the resource from cache, and also send an resource not found
+			// error to indicate resource removed.
 			delete(c.ldsCache, name)
+			c.ldsMD[name] = UpdateMetadata{Status: ServiceStatusNotExist}
```

Comment:
When does this entry get removed from `ldsMD` at all? When the watch is cancelled? Is that code coming in later PRs?

Reply:
It will be removed when watch is canceled.
It's done in watchers.go, line 198, `delete(mds, resourceName)`.